libraries/ms-common/test/test_pwm.c

```diff
@@ -1,25 +1,67 @@
 #include "pwm.h"
 
+#include <stdint.h>
+
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt.h"
+#include "interrupt_def.h"
+#include "log.h"
 #include "test_helpers.h"
 #include "unity.h"
 
 #define PERIOD_MS 100
+#define WAIT_CYCLES 5
+
+static volatile uint16_t s_counter = 0;
 
-void setup_test(void) {}
+static void prv_pwm_test(const GPIOAddress *address, void *context) {
+  s_counter++;
+}
+
+void setup_test(void) {
+  interrupt_init();
+  gpio_init();
+  gpio_it_init();
+  s_counter = 0;
+}
 
 void teardown_test(void) {}
 
 void test_pwm_guards(void) {
-  TEST_ASSERT_EQUAL(STATUS_CODE_UNINITIALIZED, pwm_set_pulse(50));
-  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_init(0));
-  TEST_ASSERT_OK(pwm_init(PERIOD_MS));
-  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_set_dc(101));
-  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_set_pulse(PERIOD_MS + 1));
+  TEST_ASSERT_EQUAL(STATUS_CODE_UNINITIALIZED, pwm_set_pulse(PWM_TIMER_3, 50));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_init(PWM_TIMER_3, 0));
+  TEST_ASSERT_OK(pwm_init(PWM_TIMER_3, PERIOD_MS));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_set_dc(PWM_TIMER_3, 101));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_set_pulse(PWM_TIMER_3, PERIOD_MS + 1));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, pwm_init(NUM_PWM_TIMERS, PERIOD_MS));
 }
 
-void test_pwm_valid_arg(void) {
-  TEST_ASSERT_OK(pwm_init(PERIOD_MS));
-  TEST_ASSERT_EQUAL(PERIOD_MS, pwm_get_period());
-  TEST_ASSERT_OK(pwm_set_dc(50));
-  TEST_ASSERT_OK(pwm_set_pulse(PERIOD_MS - 1));
+void test_pwm_output(void) {
+  TEST_ASSERT_OK(pwm_init(PWM_TIMER_3, PERIOD_MS));
+  TEST_ASSERT_EQUAL(PERIOD_MS, pwm_get_period(PWM_TIMER_3));
+  TEST_ASSERT_OK(pwm_set_pulse(PWM_TIMER_3, PERIOD_MS - 1));
+  TEST_ASSERT_OK(pwm_set_dc(PWM_TIMER_3, 50));
+
+  const GPIOAddress addr = {
+    .port = GPIO_PORT_A,  //
+    .pin = 6,             //
+  };
+  const GPIOSettings settings = {
+    .direction = GPIO_DIR_OUT,
+    .state = GPIO_STATE_HIGH,
+    .resistor = GPIO_RES_PULLUP,
+    .alt_function = GPIO_ALTFN_1,
```

Comment:
Might be helpful to start defining alternate function macros for peripherals?

Reply:
I agree with this sentiment. I'd like to do this in a separate PR as it cover PWM, SPI, I2C, UART, etc.

Some pins support multiple alternate functions of the same type which might make things somewhat ugly. I'll try to come up with a nice way of doing this. A possible API I was considering for PWM was:

```c
#define PWM_GET_TIMER(GPIOAddress)
#define PWM_GET_ALTFN(GPIOAddress, PWMTimer)
```

Based on [the datasheet](http://www.st.com/content/ccc/resource/technical/document/datasheet/cd/46/43/83/22/d3/40/c8/DM00090510.pdf/files/DM00090510.pdf/jcr:content/translations/en.DM00090510.pdf) it looks like some pins support multiple timers so I guess I'll just pick the first one?

Do you have any thoughts/opinions on how to do this?

Reply:
Hmm, I'm a little wary of selecting a timer based on the pin - if we ever want to use different PWMs for two pins on the same timer, we have no way of detecting (or alerting) that there's a collision.

Maybe something like `retarget_cfg.h`? I think making it obvious what timers can be used per pin (or vice versa) and using a compile-time assert or something to ensure the combination is valid seems like a decent option. Maybe it's overkill?